_artifactcache.py: make pull/push errors retryable

abderrahim opened the pull request:
This does two things:
- catch the correct errors in ArtifactCache
- make errors raised by network operations temporary

Fixes #1476

nanonyme commented:
While this seemed to help, this is insufficient. Same problem comes up when pushing.

nanonyme commented:
In bst1 we specifically handled UNAVAILABLE and ABORTED as retryable, nothing else, see https://github.com/apache/buildstream/commit/4f68663ee818b0a35dc59992edfb59292740d23f

nanonyme commented:
https://github.com/apache/buildstream/blob/a935429a8089b12e46a0257372efedd4fa13dbb7/src/buildstream/_artifactcache.py#L410-L415 and https://github.com/apache/buildstream/blob/a935429a8089b12e46a0257372efedd4fa13dbb7/src/buildstream/_artifactcache.py#L448-L452 probably need same fix

juergbi commented:
> probably need same fix

Yes, let's cover push as well. And I expect the same issue to exist in `_sourcecache.py`.

A possible alternative would be to enable gRPC's built-in retry mechanism for the `UNAVAILABLE` status code using `grpc.service_config` options when creating the gRPC channel. This would also cover remote execution. However, job-based retry may still make sense for other status codes.

nanonyme commented:
I like the service config idea. It probably does more granular retries than this one and covers all codepaths

nanonyme commented:
Note though https://github.com/grpc/grpc/issues/25003 may be an issue when using grpc.service_config

nanonyme commented on src/buildstream/_artifactcache.py:
```diff
@@ -410,7 +414,9 @@ def _push_artifact_blobs(self, artifact, artifact_digest, remote):
                 raise ArtifactError(
-                    "Failed to push artifact blobs with status {}: {}".format(e.code().name, e.details())
+                    "Failed to push artifact blobs with status {}: {}".format(
+                        e.code().name, e.details(), temporary=True
```
mypy test is failing because temporary is used as parameter to format rather than ArtifactError

nanonyme commented on src/buildstream/_artifactcache.py:
```diff
@@ -448,7 +454,9 @@ def _push_artifact_proto(self, element, artifact, artifact_digest, remote):
                 raise ArtifactError(
-                    "Error checking artifact cache with status {}: {}".format(e.code().name, e.details())
+                    "Error checking artifact cache with status {}: {}".format(
+                        e.code().name, e.details(), temporary=True
```
Same error here, parenthesis in wrong place

nanonyme commented:
I don't understand why the docs test fails but it doesn't seem related.

abderrahim commented:
The failures look like the ones fixed by https://gitlab.com/BuildGrid/buildbox/buildbox-fuse/-/merge_requests/39
I submitted https://gitlab.com/BuildStream/buildstream-docker-images/-/merge_requests/188 to update the docker image.

abderrahim committed:
_artifactcache.py: catch the correct errors

The artifact cache shouldn't care about grpc and should catch the
exceptions raised by lower level components (CAS and Asset cache).

abderrahim committed:
_artifactcache.py: make errors raised by network operations temporary

Fixes #1476

abderrahim commented:
This should be ready now

nanonyme commented:
Looks good to me. There were definitely retries happening with these changes in freedesktop-sdk. Tests seem to possibly be failing because local buildbox-casd is malfunctioning.

nanonyme commented:
I just noticed there's error like `/home/testuser/buildstream/.tox/py310-nocover/lib/python3.10/site-packages/_pytest/threadexception.py:75: PytestUnhandledThreadExceptionWarning: Exception in thread Thread-2369` in plugins test. This is really bad. sounds like signaling in error cases is broken.

abderrahim committed:
.github/common.env: Update Docker images

This provides buildbox-casd 0.0.54

abderrahim commented:
This should be ready now. I updated the CI image to include buildbox-casd 0.0.54 which should fix the test failures we were seeing.

juergbi reviewed (approved):
The error type changes make sense to me.

Marking all these errors as temporary may be overkill as we may be able to recognize permanent errors where we want to skip the retry. When switching to use buildbox-casd as proxy also for the Remote Asset API, retry support in BuildStream might even become redundant. However, this branch is useful as is and I think we should merge it. We can merge further improvements later.

Similar changes should be applied to `_elementsourcescache.py` and `_sourcecache.py` but that can also be done in a follow-up.

nanonyme commented:
@juergbi so is this ok to merge as is? We're currently unable to use CAS with BuildStream master at all until this is merged.

abderrahim commented:
I rebased this branch on master, so it should be mergeable now. The only test failure is the well known #1368.
